contracts/src/ERC20Permit.sol

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.20;
+
+import {IERC20} from "./interfaces/IERC20.sol";
+import {IERC20Permit} from "./interfaces/IERC20Permit.sol";
+
+import {ERC20} from "./ERC20.sol";
+import {Ownable} from "openzeppelin/access/Ownable.sol";
+
+contract ERC20Permit is IERC20, IERC20Permit, ERC20, Ownable {
+    error PermitExpired();
+    error InvalidS();
+    error InvalidV();
+    error InvalidSignature();
+
+    bytes32 public immutable DOMAIN_SEPARATOR;
+
+    string private constant EIP191_PREFIX_FOR_EIP712_STRUCTURED_DATA = "\x19\x01";
+
+    // keccak256('EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)')
+    bytes32 private constant DOMAIN_TYPE_SIGNATURE_HASH =
+        bytes32(0x8b73c3c69bb8fe3d512ecc4cf759cc79239f7b179b0ffacaa9a75d522b39400f);
+
+    // keccak256('Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)')
+    bytes32 private constant PERMIT_SIGNATURE_HASH =
+        bytes32(0x6e71edae12b1b97f4d1f60370fef10105fa2faae0126114a169c64845d6126c9);
+
+    mapping(address => uint256) public nonces;
+
+    constructor(string memory name_, string memory symbol_, uint8 decimals_) ERC20(name_, symbol_, decimals_) {
```

Comment:
Add an `address immutable owner` variable, and initialize it with `owner = msg.sender`. Similar to the pattern used in the constructor for `Agent.sol`.

Then you can also introduce a modifier similar to the following to implement owner checks:
https://github.com/Snowfork/snowbridge/blob/60a82e0d6bc6ff214124e8aa2f29d3ec566bc0e5/contracts/src/Gateway.sol#L67